### ISSUE TITLE
Crashes on Windows 10

### ISSUE THREAD

**Issue by Zhniing** [neovide/neovide#2299]
**Describe the bug**
Neovide crashes after opening an empty buffer and typing something.

Also try to open a file (about 200 lines), and scroll the buffer via the mouse wheel. Crashes as well.

**To Reproduce**
1. Run `neovide --log -- --clean` in PowerShell
2. Type something
3. Neovide freezes 1 or 2 seconds
4. Neovide crashes (window closed)

**Expected behavior**
Don't crash.

**Screenshots**
I captured a video (1.84MB in `.mp4`) of the whole process but, seems unable to upload.

**Desktop:**
 - OS: Windows 10 22H2
 - Neovide Version: build from [3e575fcd](https://github.com/neovide/neovide/commit/3e575fcd21863fdaa1fd14d23db649738694af97) (via `cargo install --path .`)
 - Neovim Version: 0.9.5 (via `scoop install neovim`)

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**

[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/14022054/neovide_rCURRENT.log)

**Additional context**
This may be a stupid issue. I tried various versions of both Neovide and Neovim, and always crashed. I just have no idea what happened. Could you help me out?

Thank you for making this impressed project!

**Comment by fredizzimo** [comment]
Do you have a neovide_backtraces.log file, in the working directory where you started neovide from?

**Comment by Zhniing** [comment]
> Do you have a neovide_backtraces.log file, in the working directory where you started neovide from?

No. I start Neovide in an empty directory. There is only one file (i.e. neovide_rCURRENT.log) generated after Neovide crashed.

**Comment by fredizzimo** [comment]
Ok, can you try first setting the environment variable `RUST_BACKTRACE=1` and then run `cargo run -- --clean --log`, that should print the backtrace to the terminal, and run a debug build.

**Comment by Zhniing** [comment]
Tried as the following process:

1. Run `$env:RUST_BACKTRACE=1` in PowerShell to set the environment variable
2. Run `cargo run -- --clean --log` and get this error:

```
error: unexpected argument '--clean' found

  tip: to pass '--clean' as a value, use '-- --clean'

Usage: neovide.exe [OPTIONS] [FILES_TO_OPEN]... [-- [NEOVIM_ARGS]...]

For more information, try '--help'.
error: process didn't exit successfully: `target\debug\neovide.exe --clean --log` (exit code: 2)
```

3. So, I remove the nvim directories manually:

```powershell
mv nvim nvim.bak
mv nvim-data nvim-data.bak
```

4. Then, run `cargo run -- --log` from the repository directory. Typed something and played in `:Tutor` for about 5 minutes. **Didn't crash!** I don't know if it's because didn't crash, no backtraces were printed to the terminal. Only printed these:

```
Finished dev [unoptimized + debuginfo] target(s) in 0.48s
     Running `target\debug\neovide.exe --log`
```

5. So, I install again with `cargo install --path .` and start Neovide again from an empty directory. **Unfortunately crashed again**

**Additionally**

My cargo was installed by rustup. My rustup was install via `scoop isntall main/rustup`

**Comment by fredizzimo** [comment]
Hm, yes, I gave the wrong command it should have been `cargo run -- --log -- --clean`

But it looks like the debug mode might not crash and only the release mode, first comment out this line, so that the output goes to the console
https://github.com/neovide/neovide/blob/3e575fcd21863fdaa1fd14d23db649738694af97/src/main.rs#L1

Then run:
`cargo run --profile profiling  -- --log -- --clean` the `profiling` profile is a bit misnamed, it's actually release with debug information enabled, so it should work the same as release. If not, you could try
`cargo run --release -- --log -- --clean`

All with the `RUST_BACKTRACE=1` environment variable set.

If that still does not help, I think a debugger is needed on your side, since I can't reproduce the bug here. If you have Visual Studio code installed, it's not too hard, especially if you have used VSCode before. You will need this https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb

And then the following launch configuration
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "lldb",
            "request": "launch",
            "args": ["--log", "--", "--clean"],
            "cargo": {
                "args": ["build", "--profile", "profiling"]
            }
        },
    ]
}
```

It's also possible to use the Neovim and the DAP protocol, but I haven't set that up that properly myself, so I can't give you instructions regarding that.

**Comment by Zhniing** [comment]
The process is as following:

#### 1. Commented out the first line of `src/main.rs` as following:

```rust
// #![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
```

#### 2. Run `cargo run --profile profiling  -- --log -- --clean the profiling`

Didn't crash and no backtraces as before (neither printed to terminal nor generated neovide_backtraces.log file)

Terminal outputs:

```
Finished profiling [optimized + debuginfo] target(s) in 7m 01s
 Running `target\profiling\neovide.exe --log -- --clean`
```

#### 3. Run `cargo run --release -- --log -- --clean`

Didn't crash and no backtraces as before (neither printed to terminal nor generated neovide_backtraces.log file)

Terminal outputs:

```
Finished release [optimized] target(s) in 9m 15s
     Running `target\release\neovide.exe --log -- --clean`
ERROR [neovide::settings::config] Could not watch config file, chances are it just doesn't exist: Input watch path is neither a file nor a directory.
```

#### 4. Debug in VSCode

Successfully start a debugger (lldb) in VSCode.

The `TERMINAL` outputs:

```
ERROR [neovide::settings::config] Could not watch config file, chances are it just doesn't exist: Input watch path is neither a file nor a directory.
```

The `OUTPUT` of `LLDB`:

```
Initial debug configuration: {
  name: 'Launch',
  type: 'lldb',
  request: 'launch',
  args: [ '--log', '--', '--clean' ],
  cargo: { args: [ 'build', '--profile', 'profiling' ] },
  __configurationTarget: 6
}
Raw artifacts:
{
  fileName: 'c:\\Users\\zhn\\Projects\\neovide\\target\\profiling\\neovide.exe',
  name: 'neovide',
  kind: 'bin'
}
Filtered artifacts: 
{
  fileName: 'c:\\Users\\zhn\\Projects\\neovide\\target\\profiling\\neovide.exe',
  name: 'neovide',
  kind: 'bin'
}
Resolved debug configuration: {
  name: 'Launch',
  type: 'lldb',
  request: 'launch',
  args: [ '--log', '--', '--clean' ],
  __configurationTarget: 6,
  relativePathBase: 'c:\\Users\\zhn\\Projects\\neovide',
  program: 'c:\\Users\\zhn\\Projects\\neovide\\target\\profiling\\neovide.exe',
  sourceLanguages: [ 'rust' ],
  _adapterSettings: {
    displayFormat: 'auto',
    showDisassembly: 'auto',
    dereferencePointers: true,
    suppressMissingSourceFiles: true,
    evaluationTimeout: 5,
    consoleMode: 'commands',
    sourceLanguages: null,
    terminalPromptClear: null,
    evaluateForHovers: true,
    commandCompletions: true,
    reproducer: false
  }
}
[adapter\codelldb\src\terminal.rs:101] winapi::um::wincon::FreeConsole() = 1
[adapter\codelldb\src\terminal.rs:102] winapi::um::wincon::AttachConsole(pid) = 1
[adapter\codelldb\src\terminal.rs:109] winapi::um::wincon::FreeConsole() = 1
```

The `DEBUG CONSOLE` outputs:

```
Console is in 'commands' mode, prefix expressions with '?'.
Launching: c:\Users\zhn\Projects\neovide\target\profiling\neovide.exe --log -- --clean
Launched process 16248
```

And the Neovide didn't crash as before, so I can't figure out the reason caused crash.

---

I'm aware that I installed the Windows 11 SDK instead of the Windows 10 SDK when I installed the rust using rustup, but my OS is Windows 10. I don't know if this matter. Just to be sure, I'm considering to reinstall the rustup (and cargo) from the offical site instead of using the scoop. Then build Neovide again.

**Comment by fredizzimo** [comment]
Ok, strange, it looks like only the installed version crashes, maybe it's running the wrong executable, and not the one you installed?

But if it's the right one, you could try to install this version, with the commented-out line. And then run `neovide -- --clean` with the `RUST_BACKTRACE=1` environment variable set. And see if that outputs something useful.

**Comment by Zhniing** [comment]
Just in case, may I ask if the install command `cargo install --path .` is proper? Are any other options recommended?

**Comment by fredizzimo** [comment]
Yes, that's the correct command to use when installing from source. 

If you run `where.exe neovide` after that it should show a path to `c:\Users\user.name\.cargo.bin\neovide.exe`, if not, then you have two installations, and need to uninstall the other one first.

**Comment by Zhniing** [comment]
This is the installed process:

```powershell
❯ cargo install --path .
  Installing neovide v0.12.1 (C:\Users\zhn\Projects\neovide)
    Updating crates.io index
   Compiling neovide v0.12.1 (C:\Users\zhn\Projects\neovide)
    Finished release [optimized] target(s) in 2m 02s
  Installing C:\Users\zhn\scoop\persist\rustup\.cargo\bin\neovide.exe
   Installed package `neovide v0.12.1 (C:\Users\zhn\Projects\neovide)` (executable `neovide.exe`)
```

This is the output of `where.exe neovide`:

```powershell
❯ where.exe neovide
C:\Users\zhn\scoop\persist\rustup\.cargo\bin\neovide.exe
C:\Users\zhn\scoop\apps\rustup\current\.cargo\bin\neovide.exe
```

The second one is just a symbol link, that is, these two `neovide.exe` are exactly the same:

```powershell
❯ lsd -l C:\Users\zhn\scoop\apps\rustup\current\.cargo
'C:\Users\zhn\scoop\apps\rustup\current\.cargo' ⇒ C:\Users\zhn\scoop\persist\rustup\.cargo
```

---

> But if it's the right one, you could try to install this version, with the commented-out line. And then run `neovide -- --clean` with the `RUST_BACKTRACE=1` environment variable set. And see if that outputs something useful.

The installation process is as described as above. I run `neovide -- --clean` with the `RUST_BACKTRACE=1`, and it crashed with the following one-line output:

```
fatal runtime error: I/O error: operation failed to complete synchronously
```

**Comment by fredizzimo** [comment]
I don't know the cause, but we have another issue with the same "I/O error: operation failed to complete synchronously" message
* https://github.com/neovide/neovide/issues/2229

You could try `cargo install --debug --path .` to see if that gives more information.

And if that does not help vscode with the following launch configuration, which should debug the installed executable
```json
{
    "version": "0.2.0",
    "configurations": [
        {
           "name": "Launch",
           "type": "lldb",
           "request": "launch",
           "program": "C:\\Users\\zhn\\scoop\\persist\\rustup\\.cargo\\bin\\neovide.exe",
           "args": ["--log", "--", "--clean"]
        }
    ]
}

**Comment by Zhniing** [comment]
I reinstalled via `cargo install --debug --path .` with the following outputs:

```powershell
  Finished dev [unoptimized + debuginfo] target(s) in 4m 15s
Installing C:\Users\zhn\scoop\persist\rustup\.cargo\bin\neovide.exe
 Installed package `neovide v0.12.1 (C:\Users\zhn\Projects\neovide)` (executable `neovide.exe`)
```

But also get the same error:

```
fatal runtime error: I/O error: operation failed to complete synchronously
```

I tried to debug the installed executable via VSCode using the launch.json you provided, but **didn't crash** again (tried many times and never crashed). It's too strange...

---

I'll try to reinstall rust and rebuild Neovide tomorrow. Thank you for your patience and the detailed guidance!

**Comment by fredizzimo** [comment]
I'm starting to run out of ideas, but updating the toolchain might help: ´rustup update`

You could also update Visual Studio, or the Microsoft c++ build tools, since the rust compiler depends on that https://learn.microsoft.com/en-us/windows/dev-environment/rust/setup#install-visual-studio-recommended-or-the-microsoft-c-build-tools. Depending on which you have installed.

Different versions of the c++ build tools inside and outside scoop might also explain the issue.

**Comment by fredizzimo** [comment]
Another thing you could try is to attach the debugger instead of launching. So start neovide normally, then run attach and select neovide from the list.
```json
{
    "version": "0.2.0",
    "configurations": [
        {
           "name": "Attach",
           "type": "lldb",
           "request": "attach",
            "pid": "${command:pickProcess}"
        }
    ]
}
```

**Comment by Zhniing** [comment]
> Another thing you could try is to attach the debugger instead of launching.

Ok, it works, and got this:

![VSCode_Attach](https://github.com/neovide/neovide/assets/16496518/568534a2-a411-4060-a4fd-a61c96d4e25a)

The partial call stack is as follows:

![VSCode_CallStack](https://github.com/neovide/neovide/assets/16496518/2f2fc2ed-eb28-4b67-a591-8095ebe5cce0)

Is there any information here

**Comment by fredizzimo** [comment]
Yes, some. It's failing to read from the stdio pipe to Neovim. And the symptoms and error message matches this bug https://github.com/rust-lang/rust/issues/98947

So, either the bug is not properly fixed, or you haven't updated your rust toolchain, what does `rustc --version` show?

Another possibility is that Neovim crashes, and triggers another similar bug. Is the neovim process still running at that point?

**Comment by fredizzimo** [comment]
Based on the hash it looks like you have the latest 1.75.0 rust version, which is bad news, since the bug should be fixed in that. But it's still best that you confirm it yourself.

And I have looked at the code, and I don't understand how it could fail. But it's possible that there's still something wrong with the rust implementation of the pipe io, which could also explain
* https://github.com/neovide/neovide/issues/2182

**Comment by fredizzimo** [comment]
Ok, I'm pretty convinced now that their fix is wrong. 

WaitforSingleObject should not be used on file handles, as for example indicated here https://stackoverflow.com/questions/775014/waitforsingleobject-on-a-file-handle

File handles are also not listed here https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject

Instead they should create an Event, and pass it to `NtReadfile`. https://learn.microsoft.com/en-us/windows/win32/devnotes/ntreadfile

Do, you think that you can compile the rust toolchain from source, if I attempt to make a fix? Or perhaps even attempt to fix their `read_synchrounous` function yourself? I can't really do any testing of myself, because I'm unable to repeat this. @Zhniing 

This wrong usage could also very well explain the exit hangs of #2182. But instead of returning too early from `WaitForSingleObject` it never returns. @sid-6581

**Comment by fredizzimo** [comment]
Thanks, that confirms that I'm probably on the right path, and that the issue is indeed caused by invalid windows api usage in the rust standard library.

I'm not sure when I can start working on the fix, unless you or @sid-6581 for example can do it. It should not take too long, but I have a few other Neovide fixes and a release I would like to get in first and also a daytime job.

**Comment by Zhniing** [comment]
Oh, I see. I would like to do something, but I haven't learned about Rust and the Windows native API yet. 😞

I am willing to assist the best of my ability, though.

Thank you for sparing your precious time!

**Comment by fredizzimo** [comment]
That's ok.

Note, that even testing the fix might be tricky, the installation from source has quite many dependencies to get right https://github.com/rust-lang/rust/blob/master/INSTALL.md. 

But I will probably start by modifying the unit tests to get them to fail for me, if I can do that, then you might not need to test until the fix ends up in the nightly rust version. But I will not be able to send a PR to rust without being able to verify both the failure and fix.

**Comment by Zhniing** [comment]
Ok, thanks, I think I would deal with the Rust source building first.

**Comment by sid-6581** [comment]
This is very puzzling. If this was a bug in rust you'd think this would happen a lot more to many more people. There seems to maybe be something specific to your system.

Have you tried nightly neovim, does that have the same issue? You can use bob to install it. If that does crash, what do you see going on in the other threads in the debugger?

**Comment by Zhniing** [comment]
> Have you tried nightly neovim, does that have the same issue?

The nightly neovim also have the same issue, the version info of neovim is as following:

```powershell
❯ vi -V1 -v
NVIM v0.10.0-dev-2331+g170c890dc
Build type: RelWithDebInfo
LuaJIT 2.1.1707061634
Compilation: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe /DWIN32 /D_WINDOWS /MD /Zi /O2 /Ob1 /GL -W3 -wd4311 -wd4146 -wd4003 -wd4715  -DUNIT_TESTING -DHAVE_UNIBILIUM -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -D_WIN32_WINNT=0x0602 -DMSWIN -DINCLUDE_GENERATED_DECLARATIONS -ID:/a/neovim/neovim/.deps/usr/include/luajit-2.1 -ID:/a/neovim/neovim/.deps/usr/include -ID:/a/neovim/neovim/build/src/nvim/auto -ID:/a/neovim/neovim/build/include -ID:/a/neovim/neovim/build/cmake.config -ID:/a/neovim/neovim/src

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "C:/Program Files (x86)/nvim/share/nvim"
```

> If that does crash, what do you see going on in the other threads in the debugger?

Is there a way to dump all the threads? I don't know how to show you the other threads. (screenshots seems too inefficient)

> If this was a bug in rust you'd think this would happen a lot more to many more people. There seems to maybe be something specific to your system.

Yes, I agree with you. I will test it on another Windows later.

**Comment by fredizzimo** [comment]
> > If this was a bug in rust you'd think this would happen a lot more to many more people. There seems to maybe be something specific to your system.
> 
> Yes, I agree with you. I will test it on another Windows later.

Not necessarily, there's a long history of problems related to the code in question.

On January 24th 2021 this was reported, with code demonstrating that read was broken. And as far as I can see that has always been the case before that
* https://github.com/rust-lang/rust/issues/81357

That was only fixed one year later on April 26th 2022:
* https://github.com/rust-lang/rust/pull/95469

This was detected as a regression on July 5th 2022. But it was not really a regression, it was just an inmproper fix
* https://github.com/rust-lang/rust/issues/98947

That was quickly fixed on July 10th:
* https://github.com/rust-lang/rust/pull/98950

But, I'm not the only one thinking that the fix is wrong https://github.com/rust-lang/rust/pull/98950#pullrequestreview-1030831527. And that is still the code used today. 

My point here is that things around this have been broken for a long time, and not affecting many people, or these issues would not stay open for this long.

I will try to create a stress test for the pipe IO, and see if we can get an easier way to reproduce these issues, and potentially report upstream.

**Comment by fredizzimo** [comment]
@Zhniing, are you still on Windows 10 22H2? Can you give the exact version number?
@thep0y, which version of Windows are you on?

Also, if someone else has the same problem which version of Windows are you using.

I'm trying to figure out if this problem is related to an exact specific Windows version.

**Comment by thep0y** [comment]
Windows 11 Pro
23H2
22631.3737

**Comment by fredizzimo** [comment]
I think that information alone rules out a problem with specific Windows versions, since both the releases and Windows versions (10 vs 11) are totally different.

**Comment by thep0y** [comment]
After I changed to a new laptop, the issue disappeared. I'm speculating without any evidence that this bug might be related to the hard drive, or at least to some piece of hardware, so not everyone would be "fortunate" enough to encounter it.
